pages/blog/json-schema-beyond-validation.md

```diff
@@ -0,0 +1,152 @@
+
+However, interpreting data definition from a JSON Schema is not always complex. For our previous example, we can almost interpret it as is. If I wanted a class in TypeScript that represented the data, it could look something like this (gonna use TS syntax as examples throughout). Notice how we use the `$id` keyword to define the name of such a class.
+
+```ts
+class SomeIdForSchema {
+  public someRequiredProperty: number;
+  public someOptionalProperty?: string;
+}
+```
+
+In theory, we use the very same validation rules and interpret them, such that the output gives us the definition of what form the data may take. Sounds easy enough right? :sweat_smile:
+
+The problem is that JSON Schema –which might seem simple on the surface— is complex underneath when you start to interpret the recursive keywords such as `not`, `if`, `then`, `else`, `allOf`, `oneOf`, etc. This causes the possibilities to be endless in terms of how the JSON Schema document can be structured (at least endless in principle).
+
+Ideally, all keyword possibilities MUST be supported with no restrictions. So let's take a look at a more complex example, that introduces the `not` keyword. We are not that interested in why one would define something like this, but merely the possibility of doing so.
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "$id": "SomeIdForSchema",
+  "additionalProperties": false,
+  "properties": {
+    "someOptionalProperty": {
+      "type": ["string", "number"]
+    } 
+  },
+  "not": {
+    "properties": {
+      "someOptionalProperty": {
+        "type": "number"
+      } 
+    }
+  }
+}
+```
+**Take a moment and think about what data would you say is valid against this schema?**
+
+In the very simplest form the JSON Schema could be converted to the following:
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "$id": "SomeIdForSchema",
+  "additionalProperties": false,
+  "properties": {
+    "someOptionalProperty": {
+      "type": "string"
+    } 
+  }
+}
+```
+Where the property `someOptionalProperty` may only be of type `string`. But, let's try and break the complex schema down step by step, to show how the validation would work against the data:
+```json
+{
+  "someOptionalProperty": "string"
+}
+```
+Notice how the `not` keyword reverses the validation result after step 5.
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#", 
+  "type": "object",               //Step 1 
+  "$id": "SomeIdForSchema", 
+  "additionalProperties": false,  //Step 2 
+  "properties": { 
+    "someOptionalProperty": {     //Step 3
+      "type": ["string", "number"]//Step 4
+    } 
+  },
+  "not": {                        //Step 6
+    "properties": {
+      "someOptionalProperty": { 
+        "type": "number"          //Step 5
```

Comment:
I miss the point why this is step 5 and not 6 😅

Reply:
Good point, I kinda jump over that point, gonna adapt the description 👍

Reply:
Added the following above `Notice how the `not` keyword reverses the validation result of step 5, which is why the inner schema is validated before the keyword itself.`

Reply:
interesting, I didn't know it is suppose to work this way